core/xml.js

```diff
@@ -334,9 +338,68 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
   if (workspace.setResizesEnabled) {
     workspace.setResizesEnabled(true);
   }
+  return newblockids;
 };
 
 /**
+ * Decode an XML DOM and create blocks on the workspace.
+ * Append the new blocks at the bottom of the existing blocks
+ * @param {!Element} xml XML DOM.
+ * @param {!Blockly.Workspace} workspace The workspace.
+ */
+Blockly.Xml.appendDomToWorkspace = function(xml, workspace) {
+  // first check if we have a workspaceSvg otherwise the block have no shape
+  // and the position does not matter
+  //if (typeof workspace.getBlocksBoundingBox() === undefined) {
+  if (workspace.hasOwnProperty('scale')) {
+    var savetab = Blockly.BlockSvg.TAB_WIDTH;
+    try {
+      Blockly.BlockSvg.TAB_WIDTH = 0;
+      var bbox = workspace.getBlocksBoundingBox();
+    } finally {
+      Blockly.BlockSvg.TAB_WIDTH = savetab;
+    } 
+    if (bbox.height > 0) { // check if any previous block
+      var offsetY = 0; // offset to add to y of the new block
+      var offsetX = 0; 
+      // check for existing open output connection of the top block
+      var blocks = workspace.getTopBlocks(true);
+      var farY = bbox.y + bbox.height;
+      var topX = bbox.x;
+      // load the new blocks into the workspace and get the ids of the new blocks
+      var newblockids = Blockly.Xml.domToWorkspace(xml,workspace);
```

Comment:
You have three calls to `domToWorkspace()`.   I think these can be reordered into one.

```
var blocks = workspace.getTopBlocks(true);  // Maybe rename priorBlocks
var bbox;
if (workspace.hasOwnProperty('scale')) {
  var savetab = Blockly.BlockSvg.TAB_WIDTH;
  try {
    Blockly.BlockSvg.TAB_WIDTH = 0;
    bbox = workspace.getBlocksBoundingBox();
  } finally {
    Blockly.BlockSvg.TAB_WIDTH = savetab;
  }
}
var newBlockIds = Blockly.Xml.domToWorkspace(xml,workspace);  // Camel case var name
if (bbox && bbox.height) {
  var offsetY = 0; // offset to add to y of the new block
  . . .
}
// No need for else statements.
return newBlockIds;  // Adding return to match domtoWorkspace()
```